optparse.3
.TH OPTPARSE 3
.\"
.\" Manual page for library function OptionParse
.\"
.\" Git id: $Id$
.\"
.SH NAME
OptionParse \- scan the arg list and return the next arg
.SH SYNOPSIS
.P
\fBint  OptionParse( int argc, char **argv, char *opts );
.br
void OptionInit();
.br
void OptionReset( char switch, errflag, externflag );
.br
extern char *OptSwitch;\fR
.SH DESCRIPTION
This suite of functions is an extension to the now-standard
.I getopt(3)
option fetching routines.
.I OptionParse()
and it's helping routines are considerably more flexible,
plus don't do such unfriendly or annoying things as writing unexpected
error messages or making your program call
.I
.I exit().
.PP
.I OptionParse()
returns an integer value containing a character.
It scans the arguement list passed in, starting
from an internal pointer into the list.
Each time it is called,
the internal pointer is reset so the next call will
scan the next arg.
.PP
An arguement is first checked to see if it begins with the
switch character.
If it does not, it is assumed to be a
non-switched arguement (eg, a file name).
A pointer to is is copied into
.I OptSwitch,
and
.I OptionParse()
returns a null character.
.PP
If it is a switched arg, the first character
following the switch is checked against the
.I opts
list.
If invalid,
.I OptSwitch
is set to point to the invalid flag and
.I errflag
is returned.
If valid,
.I OptionParse()
then checks to see if there should be an extended
arguement.
If there should not be,
.I OptSwitch
is set to the null pointer
and the switch character is returned.
.pp
If there should be an extended arg, it is checked for in two
places: first, following the switch character in the same
arguement.
If not found there, the next arguement is examined.
If not found, 
.I OptSwitch
is set to the null pointer
and the switch character is returned.
.PP
When the end of the arguement list is found,
.I OptionParse()
returns
.I EOF
as defined in
.I stdio.h.
.PP
Format of 
.I opts:
all of the valid switches
lumped together into a single string.
Switches which have
extended arguements must be followed by the extended arg
flag, ':' for default (the extended arg flag may be reset by
calling optargs()).
Thus "abc:d" would mean the switches
-a, -b, -c, and -d are all legal.
In addition, the -c
switch may have some sort of trailing parameter.
More than one switch may be concatenated into a single arguement.
Thus
.I OptionParse()
would treat
.I "-a -b"
and
.I -ab
as identical.
It would also treat
.I "-c fred"
and
.I -cfred
as identical, ie, switch
.I c
was found with the extended arguement
.I fred.
.PP
.I OptionInit()
is a way of forcing
.I OptionParse()
to reset to the beginning of the arguement list.
The next time 
.I OptionParse()
is called, it will return as if that were the first time it had been called.
Useful if you want to re-scan an arguement list,
possibly with different switch characters.
.PP
One can use
.I OptionReset()
to redefine the characters used to indicate a switch (default is '-'),
the character used to indicate an error (default is '?'),
and the character in
.I opts
(on the
.I OptionParse()
call) that indicates an extended arguement.
In this way one could handle switches that began with '+'
or use '-:' as a switch.
.PP
.I OptionReset()
does not force an
.I OptionInit().
Any character(s) in the
.I OptionReset()
call which are null are reset to the default value(s).
.PP
The following code fragment,
taken from the AT&T manual on
.I getopt(),
shows how one might process the arguements for a command
that can take the mutually exclusive options
\fBa\fR and \fBb\fR, and the options \fBf\fR and \fBo\fR,
both of which require extended arguements.
.sp
.li
main( int argc, char argv )
{
  int sw_char ;
  extern int  OptionParse() ;
  extern char  OptSwitch ;
  .
  .
  .
  while ( ( EOF != ( sw_char = OptionParse( argc, argv ) ) )
    switch ( sw_char )
    {
      case 'a':
        if ( bflag )
          ++errflag ;  
        else
          aflag ;
        break ;
      case 'b':
        if ( aflag )
          ++errflag ;  
        else
          bflag ;
        break ;
      case 'f':
        infile = OptSwitch ;
        break ;
      case 'o':
        outfile = OptSwitch ;
        break ;
      case '?':
        errflag++ ;
        illflag( "Illegal flag", *OptSwitch ) ;
      default:
        errflag++ ;
        illflag( "OptionParse failed on character", (char) sw_char ) ;
        break ;
    }
    .
    .
    .
}
.fi
.PP
Needless to say, the \fBdefault\fR case above should never occur!
.SH WARNINGS
.PP
These routines all use
.I stdio.h.
If you don't want that, you're out of luck.
.SH "SEE ALSO"
getopt(3)
.SH DIAGNOSTICS
None.
.SH WARNINGS
If you don't check the return value properly you can get yourself
into an infinite loop.
.SH "COPYRIGHT STATUS"
The OptionParse function and this manual page are
Copyright 1989 Steven C. Simmons.
They may be freely distributed and used as you please
provided this notice remains intact.
.SH AUTHOR
scs@di.org
.SH BUGS
None known.
Please report any strange behavior to the author.


optparse.c
/*
 * OptionParse() -- get command line switches according to
 *   a user-defined script
 *
 * Purpose
 *
 * This is a suite of functions which, taken together, gives
 * the user very tight control over command line parsing.
 * It consists of three entry points and a global variable:
 *
 *   OptionParse() -- get the "next" arguement
 *   OptionInit() -- can be called to "rewind" the arguement list
 *   OptionReset() -- allows the user to redefine the switches
 *   OptionSwitch -- extended arguement associated with a switch
 *
 * See the individual routines below for detailed descriptions.
 */

#ifdef  TEST
#define MAIN
#endif // of ifdef TEST

#ifndef lib
static char    gitid[] = "$Id$";
#pragma unused(gitid)
#endif // if ifndef lib

#include    <stdio.h>
#include    <stdlib.h>
#include    <ctype.h>

#include    "isdebug.h"

/*    Functions from standard system libraries    */

extern char* strchr();     // scan a string for a character

/*    Module-wide variables    */

#ifdef DEBUGGING
#define MODULE_NAME "optparse"
#endif
static int  argnum = 1;     // current location in argv
static int  argchar = 1;    // which character in arg we're looking at
static char swflag = '-';   // what character delimits a switch
static char extarg = ':';   // flag for extended argument wanted
static char errflg = '?';   // what to return on a non-existant switch
static char dummy[ 2 ];     // buffer to build error returns

/*    Global variables    */

const char* OptionSwitch = NULL;  // pointer to return an extended arguement

/*
 * OptionParse -- scan the arg list and return next arguement
 *
 * parameters
 *   argc -- maximum number of arguements
 *   argv -- array of argc pointers to arguements
 *   opts -- list of valid switches
 *
 * return value (int)
 *   null character if no switch is on the arguement, and
 *     OptionSwitch points to the arg;
 *   valid character if found behind switch, and OptionSwitch
 *     points to extended if requested (NULL if not or
 *     if there is none);
 *   errflg if illegal character found behind switch and
 *     OptionSwitch points to illegal character;
 *   EOF if no args to process.
 *
 * This function scans the arguement list passed in, starting
 * from an internal pointer into the list. Each time it is
 * called, the internal pointer is reset so the next call will
 * scan the next arg.
 *
 * An arguement is first checked to see if it begins with the
 * switch character. If it does not, it is assumed to be a
 * non-switched arguement (eg, a file name). A pointer to is
 * is copied into OptionSwitch, and we return a null character.
 *
 * If it is a switched arg, we return check the first character
 * following the switch against the opts list. If invalid, we
 * set OptionSwitch to point to it and return errflag.
 *
 * If valid, we then check to see if there should be an extended
 * arguement. If there should not be, we set OptionSwitch to NULL
 * and return the switch character.
 *
 * If there should be an extended arg, we check for it in two
 * places: first, following the switch character in the same
 * arguement. If not found there, we look for the next arg. If
 * not found, set OptionSwitch NULL. Return the switch character.
 *
 * Format of valid switches string -- all of the valid switches
 * lumped together into a single string. Switches which have
 * extended arguements must be followed by the extended arg
 * flag, ':' for default (the extended arg flag may be reset by
 * calling optargs()). Thus "abc:d" would mean the switches
 * -a, -b, -c, and -d are all legal. In addition, the -c
 * switch may have some sort of trailing parameter.
 */

int    OptionParse( const int argc, const char** const argv, const char* const opts ) {
    DEBUGTRACE("OptionParse")
    register char    c;    // General char holder
    register char*   cp;   // Pointer into options list for this switch

    OptionSwitch = NULL;

    /*
     * If we're just starting to look at this list ( argchar == 1 )
     * we just check two things -- end of arg list, and switch presence
     */
    
    if ( argchar == 1 ) {
        if ( argnum >= argc ) // At end of arg list?
            return EOF;
        if ( argv[ argnum ][ 0 ] != swflag ) { // No switch flag on arg
            OptionSwitch = argv[ argnum++ ];
            return '\0';
        } else if ( argv[ argnum ][ 1 ] == '\0' ) {
            // No arg following switch char, return switch as arg
            dummy[ 0 ] = argv[ argnum++ ][ 0 ];
            OptionSwitch = (const char*) dummy;
            return '\0';
        }
    }

    /*
     * When we're past the first character, it means we are
     * seeing multiple characters under a single switch, eg, '-TagP'.
     * This is the equivalent of '-T -a -g -P', unless of course
     * extended processing is desired. Thus, if 'T' was a solo
     * switch but 'a' required extended processing, '-TagP' would
     * be equivalent to '-T -a gP', where 'gP' is the extended arg
     * of 'a'.
     */
    
    c = argv[ argnum ][ argchar ];
    // Check to see if this is a good switch
    if ( ( cp = strchr( opts, c ) ) == NULL ) {
        // Bad switch.
        if ( argv[ argnum ][ ++argchar ] == '\0' ) {
            // If end of arg, reset to look at next arg
            argnum++;
            argchar = 1;
        }
        // Build the error return and return it
        dummy[ 0 ] = c;
        OptionSwitch = (const char*) dummy;
        return errflg;
    }

    /*
     * Check for extended arg with this switch. Check both
     * formats, either '-S extarg' (two args) or '-Sextarg' (one arg)
     */
    
    if ( *++cp == extarg ) {
        if ( argv[ argnum ][ argchar + 1 ] != '\0' ) {
            OptionSwitch = &argv[ argnum++ ][ argchar + 1 ];
        } else if ( ++argnum < argc ) { // Found extended
            OptionSwitch = argv[ argnum++ ];
        }
        argchar = 1; // Start at beginning of next arg
    } else if ( argv[ argnum ][ ++argchar ] == '\0' ) {
        // Arg complete -- reset for next arg
        argchar = 1;
        argnum++;
    }
    return c;
}


/*
 * OptionInit -- reset option scanner to start of arglist
 *
 * parameters
 *   none
 *
 * return value
 *   none
 *
 * This function will reset the various internal counters and
 * pointer such that the next time OptionParse is called, scanning
 * will start at the beginning of the list.
 */

void    OptionInit( void ) {
    DEBUGTRACE("OptionInit")
    argchar = argnum = 1;
    OptionSwitch = NULL;
}



/*
 * OptionReset -- redefine the various options flags and reinit OptionParse
 *
 * parameters
 *   switchchar - the character that will become the switch
 *   errorflag -- the character that will be returned for errors
 *   externflag -- the character that will indicate extended args
 *
 * return value
 *   none
 *
 * Check the incoming parameters. If any are null, set that
 * particular flag to the default value. Force incoming flags
 * to be ascii.
 */

void    OptionReset( const char switchchar, const char errorflag, const char externflag ) {
    DEBUGTRACE("OptionReset")
    swflag = ( switchchar == '\0' ) ? '-' : toascii( switchchar );
    errflg = ( errorflag == '\0' ) ? '-' : toascii( errorflag );
    extarg = ( externflag == '\0' ) ? '-' : toascii( externflag );
}

#ifdef TEST
/*
 * This program allows you to call OptionParse with various
 * command switches, and find out what OptionParse says the
 * various switches are.
 */

int main( int argc, char** argv ) {
    int    c;
    int    i;
    const char* const    opts = "abc:DefF:G";

    i = 1;
    (void) printf( "Option list is \"%s\"\n", opts );
    while ( EOF != ( c = OptionParse( argc, argv, opts ) ) ) {
        if ( c == '\0' ) {
            (void) printf( "Pass %d -- no arg, OptionSwitch is ", i );
            if ( OptionSwitch == NULL ) {
                (void) printf( "NULL\n" );
            } else {
                (void) printf( "\"%s\"\n", OptionSwitch );
            }
        } else if ( c == '?' ) {
            (void) printf( "Pass %d -- err arg: ", i );
            if ( OptionSwitch == NULL ) {
                (void) printf( "NULL\n" );
            } else {
                (void) printf( "\"%s\"\n", OptionSwitch );
            }
        } else {
            (void) printf( "Pass %d -- Got option '%c',", i, c );
            (void) printf( " got OptionSwitch arg " );
            if ( OptionSwitch == NULL ) {
                (void) printf( "NULL\n" );
            } else {
                (void) printf( "\"%s\"\n", OptionSwitch );
            }
        }
        i++;
    }
    return 0;
}
#endif // of ifdef TEST
